// @cloudbase/node-sdk/lib/utils/metadata.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const axios_1 = __importDefault(require("axios"));
exports.kMetadataBaseUrl = 'http://metadata.tencentyun.com';
var kMetadataVersions;
(function (kMetadataVersions) {
    kMetadataVersions["v20170919"] = "2017-09-19";
    kMetadataVersions["v1.0"] = "1.0";
    kMetadataVersions["latest"] = "latest";
})(kMetadataVersions = exports.kMetadataVersions || (exports.kMetadataVersions = {}));
function isAppId(appIdStr) {
    return /^[1-9][0-9]{4,64}$/gim.test(appIdStr);
}
exports.isAppId = isAppId;
async function lookup(path, options = {}) {
    const url = `${exports.kMetadataBaseUrl}/${kMetadataVersions.latest}/${path}`;
    const resp = await axios_1.default.get(url, options);
    if (resp.status === 200) {
        return resp.data;
    }
    else {
        throw new Error(`[ERROR] GET ${url} status: ${resp.status}`);
    }
}
exports.lookup = lookup;
const metadataCache = {
    appId: undefined
};
/**
 * lookupAppId - è¯¥æ–¹æ³•ä¸»è¦ç”¨äºåˆ¤æ–­æ˜¯å¦åœ¨äº‘ä¸Šç¯å¢ƒ
 * @returns
 */
async function lookupAppId() {
    if (metadataCache.appId === undefined) {
        metadataCache.appId = '';
        try {
            // åªæœ‰é¦–æ¬¡ä¼šè¯·æ±‚ä¸”è¦æ±‚å¿«é€Ÿè¿”å›ï¼Œè¶…æ—¶æ—¶é—´å¾ˆçŸ­ï¼ŒDNSæ— æ³•è§£æå°†ä¼šè¶…æ—¶è¿”å›
            // åœ¨äº‘ç¯å¢ƒä¸­ï¼Œè¿™ä¸ªæ—¶é—´é€šå¸¸åœ¨ 10ms å†…ï¼Œéƒ¨åˆ†è€—æ—¶é•¿ï¼ˆ30+msï¼‰çš„æƒ…å†µæ˜¯ DNS è§£æè€—æ—¶é•¿ï¼ˆ27+msï¼‰
            const appId = await lookup('meta-data/app-id', { timeout: 30 });
            if (isAppId(appId)) {
                metadataCache.appId = appId;
            }
        }
        catch (e) {
            // ignore
        }
    }
    return metadataCache.appId || '';
}
exports.lookupAppId = lookupAppId;
